design/ceph/cephx-keyring-rotation.md

```diff
@@ -0,0 +1,438 @@
+    intended use case).
+    If this is set to less than or equal to the current key generation, keys are not rotated.
+
+### Enabling local key rotations
+
+Ceph developers have suggested rotation on every Ceph version update. This corresponds to a rotation
+each time the administrator updates the CephCluster's Ceph image. This also aligns with when Rook
+knows Ceph daemons are going to be restarted and when Rook can reasonably assume the administrator
+has a maintenance window. Rotation at this periodicity will have no additional impact on cluster
+connectivity or performance, so this periodicity will be used as the suggested periodicity option,
+and one-time updates will be allowed as well.
+
+For ease of configuration, the option for rotating local keys will be present on only the
+CephCluster CR. Any child CRs (e.g., CephFilesystem) dependent on the CephCluster will inherit the
+local key rotation config from the corresponding CephCluster. This allows administrators to enable
+key rotation selectively for specific CephClusters while also keeping the UX simple.
+
+```yaml
+spec:
+  # ...
+  security:
+    cephx:
+      local:
+        keyRotationPolicy: Disabled | WithCephVersionUpgrade | KeyGeneration
+        keyGeneration: <int> # used with KeyGeneration
+      csi: {} # discussed more later
+      # (room for future spec.security.cephx options)
+    # (room for future spec.security options)
+  # ...
+```
+
+List of Ceph daemons which have local keys that can be rotated automatically:
+
+- mon - all share the same key
+- mgr
+- osd
+- mds
+- rgw
+- nfs
+- rbdmirror - has 2 keys: one local (will auto-rotate) and one mirror peer (will not auto-rotate)
+- cephfsmirror - same considerations as rbdmirror
+- log-rotator
+- crash-collector
+
+The Ceph admin key Rook uses to run Ceph commands will also be rotated automatically. Care should be
+taken to ensure admin key rotation doesn't block rotation of other keys.
+
+Rook considered using an operator-level global config option
+`ROOK_ROTATE_LOCAL_CEPHX_KEYS_OLDER_THAN`, but this does not allow controllers to get reconciliation
+events when the config is modified. Associating the config with CephCluster allows controllers to
+reconcile as needed if/when the user modifies the configuration(s).
+
+### Enabling CSI key rotation
+
+Ceph-CSI deployments (if managed by Rook) are managed in the operator namespace, but keys are
+created on a per-CephCluster basis. Thus, a CephCluster configuration option (like above) is most
+appropriate. CSI key rotations may require manual administrator action to reboot or drain/undrain
+nodes to PVCs, so Rook must allow users to initiate one-time key rotations.
+
+```yaml
+spec:
+  # ...
+  security:
+    cephx:
+      local: {} # discussed above
+      csi:
+        keyRotationPolicy: Disabled | KeyGeneration
+        keyGeneration: <int> # used with KeyGeneration
+      # (room for future spec.security.cephx options)
```

Comment:
@BlaineEXE Is these are the Keys Rook generating today for cephFS and RBD?

Reply:
I don't understand the question

Reply:
This is only for these keys right https://github.com/rook/rook/blob/cf5bbd04aea0854237a9252838d142d77a0ac7a9/pkg/operator/ceph/csi/secrets.go#L30-L43 ?

Reply:
sounds good.